### PR TITLE
Add sections on Spec2 + Spec4-SMT

### PR DESCRIPTION
**Pull request by thpani**
Add sections on `Spec2` and the SMT-Spec to the report

### CODE REVIEW

File: reports/spec2.tex [hunk 1]
```diff
@@ -1,5 +1,91 @@
...
+\SpecOne{} while eliminating recursion. In \SpecOne{}, (mutually) recursive
+operators model key aspects of protocol behavior, such as the block tree and
+block justification. Apalache does not natively support recursive
+operators\footnote{\url{https://apalache-mc.org/docs/apalache/principles/recursive.html}},
+thus \SpecOne{} cannot not be directly model-checked.
+
+To resolve this, we reformulate \SpecOne{} into \SpecTwo{}, by substituting
+(mutually) recursive constructs with bounded
+\texttt{fold}~operations\footnote{In functional programming, \texttt{fold} is a
+  higher-order function that uses of a given combining operation to build a
```
Comment (Kukovec) [inline reports/spec2.tex:20]: > that uses of

feels like a noun went missing here

**Review by Kukovec — APPROVED** [review]
lgtm

**Comment by konnov** [comment]
sent comments offchain :)

File: reports/spec2.tex (hunk 1 above)
Comment (thpani) [reply reports/spec2.tex]: Fixed in 1085981

File: reports/spec2.tex [hunk 2]
```diff
@@ -73,19 +76,104 @@ \subsection{An Optimization: Flattening Nested Folds}
...
-specifications from the outset can be more effective. When relying on
-autogenerated specifications, it becomes crucial to conduct manual optimizations
-to ensure efficiency.
+To further improve our confidence in the correctness of this optimization, we
+could produce a proof in TLAPS or run Apalache to show functional equivalence.
+
+\subsection{Checking the Specification}
+
+We can query the specification for reachable protocol states using Apalache.
+For example, we can check if a finalized checkpoint exists by writing an
```
Comment (Kukovec) [inline reports/spec2.tex:95]:
```suggestion
For example, we can check if a nontrivial finalized checkpoint exists by writing an
```

File: reports/spec2.tex [hunk 3]
```diff
@@ -73,19 +76,104 @@ \subsection{An Optimization: Flattening Nested Folds}
...
+\end{table}
+
+We can see that the solver can handle the single chain block graph (where
+\textit{AccountableSafety} trivially holds due to absence of conflicting
+blocks), but struggles with the more complex scenarios even when given a fixed
+block graph. This suggests that the complexity inherent in the specification is
+due to the high combinatorial complexity of voting scenarios, rather than just
+the block graph.
+
+\subsection{Discussion}~\footnote{\color{red}TODO: perhaps move this to a general
```
Comment (Kukovec) [inline reports/spec2.tex:171]: My 2c: not against it, per se, but it's fine here imo.

**Review by Kukovec — APPROVED** [review]
👍

File: reports/spec2.tex [hunk 4]
```diff
@@ -1,5 +1,179 @@
...
+discussion section?} Applying translation rules to obtain checkable
+specifications from other artifacts can provide a useful starting point.
+However, such translation are also likely to introduce unwanted
```
Comment (konnov) [inline reports/spec2.tex:174]:
```suggestion
However, such translations are also likely to introduce unwanted
```

**Review by konnov — APPROVED** [review]
Looks good for merging!

**Pull request merged**